undofx/src/main/java/org/fxmisc/undo/impl/UndoManagerImpl.java

```diff
@@ -209,7 +229,7 @@ private void addChange(C change) {
     @SuppressWarnings("unchecked")
     private C[] merge(C c1, C c2) {
         Optional<C> merged = merge.apply(c1, c2);
-        if(merged.isPresent()) {
+        if(merged.isPresent() && isNotIdentity.test(merged.get())) {
             return (C[]) new Object[] { merged.get() };
         } else {
             return (C[]) new Object[] { c1, c2 };
```

Comment:
> Not sure where/how `Either<Barrier, C>` would be used.

I just meant that the type of queue would change from 

```java
ChangeQueue<C> queue;
```

to

```java
ChangeQueue<Either<Barrier, C>> queue;
```

Seems like you are proposing to change it to

```java
ChangeQueue<Either<C, C>> queue;
```

which would allow us to preserve the information about which changes are mergeable as well.

Neither of these might be necessary for now, after all. (See below.)

> ````
> .... c1 |~ // prevent merge called
> .... c1 c2 c3 ~ // more changes pushed
> .... c1 ~ c2 c3 // undo called twice
> ````
> When a new change (`c4`) is emitted, should `c1` merge with `c4`? It would seem the latest release would merge them. Is that supposed to happen?

It wouldn't, thanks to [this line](https://github.com/TomasMikula/UndoFX/blob/3d5555cfd84e3740e4cee004ada51abeebb93789/undofx/src/main/java/org/fxmisc/undo/impl/UndoManagerImpl.java#L93) that sets `canMerge` to `false`. It is questionable whether this the right thing to do (one might want to prevent merge at this point only if `preventMerge` was called after `c1`, but that information is now lost), but no one has complained yet :)

If we, for now, allow ourselves the same amount sloppiness, we could also just set `canMerge` to `false` after annihilation, and do "the right thing" in a separate PR.